transport-native-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollChannel.java

```diff
@@ -214,6 +209,72 @@ protected static void checkResolvable(InetSocketAddress addr) {
         }
     }
 
+    /**
+     * Read bytes into the given {@link ByteBuf} and return the amount.
+     */
+    protected final int doReadBytes(ByteBuf byteBuf) throws Exception {
+        int writerIndex = byteBuf.writerIndex();
+        int localReadAmount;
+        if (byteBuf.hasMemoryAddress()) {
+            localReadAmount = Native.readAddress(fd, byteBuf.memoryAddress(), writerIndex, byteBuf.capacity());
+        } else {
+            ByteBuffer buf = byteBuf.internalNioBuffer(writerIndex, byteBuf.writableBytes());
+            localReadAmount = Native.read(fd, buf, buf.position(), buf.limit());
+        }
+        if (localReadAmount > 0) {
+            byteBuf.writerIndex(writerIndex + localReadAmount);
+        }
+        return localReadAmount;
+    }
+
+    protected final int doWriteBytes(ByteBuf buf) throws Exception {
+        int readableBytes = buf.readableBytes();
+        int writtenBytes = 0;
+        if (buf.hasMemoryAddress()) {
+            long memoryAddress = buf.memoryAddress();
+            int readerIndex = buf.readerIndex();
+            int writerIndex = buf.writerIndex();
+            for (;;) {
+                int localFlushedAmount = Native.writeAddress(fd, memoryAddress, readerIndex, writerIndex);
+                if (localFlushedAmount > 0) {
+                    writtenBytes += localFlushedAmount;
+                    if (writtenBytes == readableBytes) {
+                        return writtenBytes;
+                    }
+                    readerIndex += localFlushedAmount;
+                } else {
+                    // Returned EAGAIN need to set EPOLLOUT
+                    setEpollOut();
+                    return writtenBytes;
+                }
+            }
+        } else {
+            ByteBuffer nioBuf;
+            if (buf.nioBufferCount() == 1) {
+                nioBuf = buf.internalNioBuffer(buf.readerIndex(), buf.readableBytes());
+            } else {
+                nioBuf = buf.nioBuffer();
```

Comment:
Has adding something like `forEach(ByteBufferProcess processor)` to the `ByteBuf` interface been discussed recently?  Where `ByteBufferProcess` would have a method which iterates over all the "internal" ByteBuffer objects. The goal of this interface would be to prevent having to always combine everything into 1 `ByteBuffer` all the time.  

Some additional complexities of this approach may include:
- checking if each `ByteBuffer` is direct or not, and then copying into direct to do native writes if necessary.
- supporting a `index` and `length` may make the interface a bit wider.